### ISSUE TITLE
console support for stdio in Windows

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2157]
|<img alt="synapsis's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| synapsis reported|
|-|-|
|Trac ID|trac#2157|
|Type|enhancement|
|Created|2006-10-07 21:44:54Z|
|Branch|https://github.com/twisted/twisted/tree/win-stdio-2157-4|

This patch allows twisted.internet.stdio to work when connected to a console, in Windows.

I have done a small refactoring in twisted.internet._pollingfile.
I think pipe support should go in twisted.internet._dumbwin32proc and the console support in a new twisted.internet._conio module.

I have also written a simple test suite.

Console support is still very limited, but now stdiodemo.py and stdin.py examples works.

In the near future I hope to have a working conch.

Attachments:

* [conio.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc1/bc1a1f5e1875e3916492b3b509f58cd420eba1d5/bc64f0619c86426c1aed3b8c87e88ba82f2bff05.patch) (15888 bytes) - added by synapsis on 2006-10-07 21:45:42Z - patch for console support
* [test_conio.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc1/bc1a1f5e1875e3916492b3b509f58cd420eba1d5/1e85e4e90fd6a79d77fbd3c939a80a45d48f6823.py) (4571 bytes) - added by synapsis on 2006-10-07 21:46:15Z - test suite for console support
* [_pollingfile.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc1/bc1a1f5e1875e3916492b3b509f58cd420eba1d5/0d162711c1b60644aa0a4c99192ec6bf2c3a163a.patch) (6767 bytes) - added by synapsis on 2006-10-22 17:25:52Z - patch _pollingfile.py, version 2
* [_pollingfile.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc1/bc1a1f5e1875e3916492b3b509f58cd420eba1d5/8fb672eedf5c9ea19fea5e5ffe7672dc119928cd.patch) (6767 bytes) - added by synapsis on 2006-10-22 17:25:58Z - patch _pollingfile.py, version 2
* [_win32stdio.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc1/bc1a1f5e1875e3916492b3b509f58cd420eba1d5/867ff4b11fa9b5132438b2981ff7f414e9983cda.patch) (7585 bytes) - added by synapsis on 2006-10-22 17:26:49Z - patch for _win32stdio.py, version 2
* [conio.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc1/bc1a1f5e1875e3916492b3b509f58cd420eba1d5/83c044a2f246e62051093254ec11eb444d1dc180.py) (9992 bytes) - added by synapsis on 2006-10-22 17:27:22Z - console support for Twisted, version 2
* [test_conio.2.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc1/bc1a1f5e1875e3916492b3b509f58cd420eba1d5/25e8bbbc36a8830eefc82402ed74ccd961b84c65.py) (7877 bytes) - added by synapsis on 2006-10-22 17:27:50Z - test suite for console support, version 2
* [win32_console_io.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc1/bc1a1f5e1875e3916492b3b509f58cd420eba1d5/c90a7410a59f9bc8420799b8ab37cf87d15075ce.patch) (33394 bytes) - added by khorn on 2009-11-10 19:18:01Z - Unified diff of above patches against r27422
* [t-internet-t-test.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc1/bc1a1f5e1875e3916492b3b509f58cd420eba1d5/eedbabfe51e240c341a1e9be69748cb0a9198e5f.patch) (19978 bytes) - added by johnnypops on 2013-04-19 04:55:07Z - _pollingfile, _dumbwin32proc, _win32stdio, test_stdio, test_pollingfile
* [win32conio.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc1/bc1a1f5e1875e3916492b3b509f58cd420eba1d5/45bda5444756a1ec752e2790393f989111456058.py) (7861 bytes) - added by johnnypops on 2013-04-19 04:56:43Z - twisted.internet.win32conio
* [test_win32conio.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc1/bc1a1f5e1875e3916492b3b509f58cd420eba1d5/6225acead9c7763d36522762a1190020f6aa111f.py) (8854 bytes) - added by johnnypops on 2013-04-19 04:57:37Z - twisted.internet.test.test_win32conio
* [windows.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc1/bc1a1f5e1875e3916492b3b509f58cd420eba1d5/91f5edf37056876e8c871021507247f60232ff61.py) (7696 bytes) - added by johnnypops on 2013-04-19 04:58:56Z - twiste.conch.windows
* [t-conch-t-python.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc1/bc1a1f5e1875e3916492b3b509f58cd420eba1d5/b683c2d57b5d19e4fb24a78b44e76ad9129fe19f.patch) (32840 bytes) - added by johnnypops on 2013-04-19 05:03:57Z - patches for cftp, conch scripts plus conch client, ssh, stdio, tap, python.win32
* [t-conch-test.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc1/bc1a1f5e1875e3916492b3b509f58cd420eba1d5/3fe32cdd03544c4fc043a9dadce0a9599e9ee9a7.patch) (23769 bytes) - added by johnnypops on 2013-04-19 05:06:07Z - lots of minor tweaks to the conch tests
* [win32conio.2.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc1/bc1a1f5e1875e3916492b3b509f58cd420eba1d5/1b246fbf4b981a58f4c71b6ed370af9c5e952967.py) (7840 bytes) - added by johnnypops on 2013-04-19 08:28:51Z - Fluffed it. Fixed version.
* [_win32stdio.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc1/bc1a1f5e1875e3916492b3b509f58cd420eba1d5/b32b534924092d1d3d4cf2be70eaf94d02a5bbc0.py) (4904 bytes) - added by johnnypops on 2013-04-19 08:29:35Z - Fluffed it. Fixed version.
* [t-internet-t-test.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc1/bc1a1f5e1875e3916492b3b509f58cd420eba1d5/352f8752e538d01ca19e984cdce28bae530ae561.patch) (19965 bytes) - added by johnnypops on 2013-04-19 10:46:42Z - Ignore the attached _win32stdio, this fixed patch includes those changes

<details><summary>Searchable metadata</summary>

```
trac-id__2157 2157
type__enhancement enhancement
reporter__synapsis synapsis
priority__normal normal
milestone__ 
branch__branches_win_stdio_2157_4 branches/win-stdio-2157-4
branch_author__synapsis__khorn__glyph__therve__thijs synapsis, khorn, glyph, therve, thijs
status__new new
resolution__None None
component__core core
keywords__windows windows
time__1160257494000000 1160257494000000
changetime__1410494892044596 1410494892044596
version__None None
owner__khorn khorn
cc__manlio_perillo@... cc__moonfallen cc__teratorn cc__strank cc__khorn cc__thijs cc__therve cc__jesstess cc__camlorn
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to synapsis|
|-|-|

Some comments on the attached code:

  * classic to new-style class conversion: good (but unrelated)
  * generalization of _PollableReadPipe and _PollableWritePipe - good
  * _PollableReadConsole - looks like a good start at the low-levels of dealing with a windows console, but the factoring needs to be cleaned up.  Dealing with the win32-specific events should not be handled by the same class which implements echoing and so forth.  As a long term goal, providing a win32 implementation of the insults ITerminalTransport/ITerminalProtocol interfaces would seem to make sense (and it may make sense to make some adjustment to these interfaces based on what win32 is capable of).  This further suggests splitting the transport logic (ie, WriteConsole calls) from the protocol logic (ie, should an input be echoed?), since insults is part of Twisted Conch, but the core of win32 conio support belongs in Twisted Core.
  * test suite - looks good.  Adding docstrings explaining the purpose of each test method would be even better.

I don't know much about the Win32 APIs involved here, so I can't comment much on the correctness of most of this code, but it looks like a nice start in the right direction.

**Comment by twisted-trac** [comment]
|<img alt="synapsis's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|synapsis commented|
|-|-|

Replying to [exarkun](#note_1):
> Some comments on the attached code:
> 
>   * classic to new-style class conversion: good (but unrelated)

Yes.
I have done some modification on _pollingfile for private use, forgetting to revert them.

Shold I remove them?

>   * generalization of _PollableReadPipe and _PollableWritePipe - good

>   * _PollableReadConsole - looks like a good start at the low-levels of dealing with a windows console, but the factoring needs to be cleaned up.  Dealing with the win32-specific events should not be handled by the same class which implements echoing and so forth.  

One possible (but complex) solution is to provide file like objects for console stdin/stdout that are POSIX compatible (like done by Cygwin).


> As a long term goal, providing a win32 implementation of the insults ITerminalTransport/ITerminalProtocol interfaces would seem to make sense (and it may make sense to make some adjustment to these interfaces based on what win32 is capable of).  

IPython contains a pyreadline package that supports ANSI terminal escape characters.

> This further suggests splitting the transport logic (ie, WriteConsole calls) from the protocol logic (ie, should an input be echoed?), since insults is part of Twisted Conch, but the core of win32 conio support belongs in Twisted Core.
>   * test suite - looks good.  Adding docstrings explaining the purpose of each test method would be even better.
> 
> I don't know much about the Win32 APIs involved here, so I can't comment much on the correctness of most of this code, but it looks like a nice start in the right direction.


Some comments of what needs to be done:

1) pipe support should (IMHO) be moved from _pollingfile to _dumbwin32proc, console support should be moved to a new _conio module

2) Since a Window process can have only one console and a console can have only on input buffer, the _PollableReadConsole should be a singleton/borg.
   The _conio module should provide globals stdin/stdout/stderr objects

3) Since in a child process can happen(?) that stdin is attached to a console and 
stdout to a pipe (or viceversa) _win32stdio should use isatty on both file descriptors

4) In _PollableReadConsole I convert '\r' to '\r\n'.
   I have done so to be able to send empty lines to a LineReceiver protocol.
   Maybe '\r' should be simply converted to '\n'?

5) I really don't know if _PollableWriteConsole can block.
   I have just copied the implementation of _PollableWritePipe, using WriteConsole instead of WriteFile

Can I implement the changes from 1)?

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|PenguinOfDoom commented|
|-|-|

Please make the IO mechanism pluggable. For example, a reactor that can wait on win32 handles (such as win32eventreactor or, potentially, iocpreactor) could turn off mouse and focus events and use WaitForMultipleObjects to find out when ReadConsole wouldn't block.

At the very least, StandardIO should not be inheriting from _PollingTimer.

Also, the standard API for closing a file handle is its loseConnection() method. I don't know why _pollingfile uses close().

**Comment by twisted-trac** [comment]
|<img alt="moonfallen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|moonfallen commented|
|-|-|

I only have a vague idea about the design details you guys are throwing around here, but I applaud this effort and I hold out hope that the win32 platform will be greatly improved by this patch.  So I'm only going to make a simple plea.  There seem to be an awful lot of modules and classes that start with underscores.  Please review the API at least once and think hard about whether so many things need to be private/untouchable, and which things can be available to applications.

Thanks.

**Comment by twisted-trac** [comment]
|<img alt="synapsis's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|synapsis commented|
|-|-|

Replying to [PenguinOfDoom](#note_3):
> Please make the IO mechanism pluggable. For example, a reactor that can wait on win32 handles (such as win32eventreactor or, potentially, iocpreactor) could turn off mouse and focus events and use WaitForMultipleObjects to find out when ReadConsole wouldn't block.
> 

This is a good idea.
I'm planning to write a file like object that allows a non blocking read from a console.

> At the very least, StandardIO should not be inheriting from _PollingTimer.
> 

I'm not sure about this.
When running the win32eventreactor we can receive notifications about input available.
But what about _PollableWriteConsole?

Moreover I really don't know if WriteConsole can block.

> Also, the standard API for closing a file handle is its loseConnection() method. I don't know why _pollingfile uses close().

**Comment by twisted-trac** [comment]
|<img alt="synapsis's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|synapsis commented|
|-|-|

This is a new set of patches for a better implementation of console support in twisted.

I have created a separate patch for _pollingfile so that this can be moved to a new ticket.

Notes:

1) I have changed _pollingfile to allow it to be reusable.
More things can be done, like moving pipe support to _dumbwin32process and refactoring _PollingTimer to a more reactor like interface

2) implementation of conio needs to be reviewed. I have to do some tests with a real POSIX implementation (on Linux).

3) support for IConsumer and IProducer (in _win32stdio) needs to be reviewed

**Comment by twisted-trac** [comment]
|<img alt="synapsis's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|synapsis commented|
|-|-|

Starting with this year I no more use Windows, so I have no more much interest in this patch.

However if this patch is still useful and there is someone that can do a review, I can try to complete it.

**Comment by twisted-trac** [comment]
|<img alt="rotund's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|rotund commented|
|-|-|

Adding "windows" keyword.

**Comment by twisted-trac** [comment]
|<img alt="czzl's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|czzl commented|
|-|-|

Duplicate of [#2135](https://github.com/twisted/twisted/issues/2135).  I kept this one open because it has a lot more stuff on it.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_13"></a>|@khorn commented|
|-|-|

Consolidated patches into a single unified diff against the current trunk (r27422).

Hopefully this will make it easier to move this forward.

With this patch, tests pass, and the stdiodemo.py example works.

The stdin.py demo does NOT work (no output), at least on my machine (WinXP).

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_14"></a>|@khorn commented|
|-|-|

Found the problem with the stdin.py demo.  It uses os.linesep as a delimiter (which on Win32 is '\r\n'), while the stdiodemo.py example uses a hardcoded '\n'.

Changing to a hardcoded '\n' in stdin.py makes it work with this patch.  I'm sure there's something deeper going on here which probably needs further investigation.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_15"></a>|@thijstriemstra commented|
|-|-|

(In [27426]) Branching to 'win-stdio-2157'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_16"></a>|@thijstriemstra commented|
|-|-|

(In [27427]) Apply win32_console_io.2.patch by khorn, refs [#2157](https://github.com/twisted/twisted/issues/2157)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_17"></a>|@thijstriemstra commented|
|-|-|

(In [27428]) Apply coding standard, refs [#2157](https://github.com/twisted/twisted/issues/2157)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_18"></a>|@thijstriemstra removed owner|
|-|-|

I slapped the patch in a branch and applied some coding standards, up for review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve commented|
|-|-|

(In [27500]) Branching to 'win-stdio-2157-2'

**Comment by twisted-trac** [comment]
|<img alt="TimAllen's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_20"></a>|TimAllen set owner to @khorn|
|-|-|

I don't have a Windows machine here to properly test this code, but here's some comments:

 1. There's a lot of "XXX" and "TODO" comments in this code. The coding standard says "Comments marked with XXX or TODO must contain a reference to the associated ticket". Hopefully the questions and unhandled conditions marked by these strings will be answered and handled before this branch is merged to trunk, and no tickets will need to be filed. :)
 1. `t/i/_pollingfile.py`, line 239 in `_PollableReadPipe.checkWork()`: Commented-out code should be fixed if it ought to work, or deleted otherwise.
 1. `t/i/_win32stdio.py`, line 34 in `_PollableReadConsole.checkWork()`: Rather than explain the magic value 2000 as "80x25", why not just write it as "80 * 25"? Also, does this break if the console is a different size (such as 80×50)?
 1. `t/i/_win32stdio.py`, line 36 in `_PollableReadConsole.checkWork()`: If you do get an unexpected `IOError`, it would be more helpful to raise it than to raise an unrelated `AssertionError`.
 1. `t/i/_win32stdio.py`, in `_PollableWriteConsole.checkWork()`: Variables `numBytesWritten` and `nBytesWritten` in the same method? There's got to be better names than these.
 1. `t/i/_win32stdio.py`, line 90: Is there a reason that `ConsoleReader` doesn't have a leading underscore, where `_PollableReadConsole` (presumably exactly the same thing but for other reactors) does not?
 1. `t/i/_win32stdio.py`, line 90: You have `ConsoleReader` for `win32eventreactor` and `_PollableReadConsole` otherwise; but `_PollableWriteConsole` works with all reactors. Why not drop `ConsoleReader` and just use `_PollableReadConsole` all the time, too?
 1. `t/i/_win32stdio.py`, line 92 in `ConsoleReader.__init__()`: This is a terrible docstring. It doesn't describe what the class represents or how it's intended to be used.
 1. `t/i/_win32stdio.py`, line 106 in `ConsoleReader.doRead()`: This looks exactly like `_PollableReadConsole.checkWork()`. Maybe `ConsoleReader` and `_PollableReadConsole` should inherit from some common base class or mixin? Otherwise, all the comments for `_PollableReadConsole.checkWork()` apply here too.
 1. `t/i/_win32stdio.py`, line 169 in `StandardIO.__init__()`: You say "for stdfd in range(0, 1, 2)", but I don't think that does what you intend. How about "for stdfd in (0, 1, 2)"?
 1. `t/i/_win32stdio.py`, line 176 in `StandardIO.__init__()`: `win32Enabled` is an unfortunate name - it's quite possible to be using Win32 with the select reactor or IOCP reactor. How about calling it `eventReactorInstalled`?
 1. `t/i/_win32stdio.py`, line 210 in `StandardIO.__init__()`: I feel it would be more readable to paste "`self._addPollableResource(self.stdin)`" into both places where it's required than to do the little if dance afterward.
 1. `t/i/conio.py`: Since this is just console IO for Windows rather than console IO in general, perhaps this module should be called `win32conio.py`?
 1. `t/t/test_conio.py`: Why is this in `twisted/test/` and not `twisted/internet/test/`, since it's testing things in `twisted/internet/`?
 1. `t/t/test_conio.py`: This test suite explodes horrible on non-Win32 platforms (or presumably Win32 platforms without the Win32 Extensions package installed) because it imports `win32console`. See `t/i/t/test_pollingfile.py` for an example of how to write platform-dependent tests that don't explode on other platforms.

I'm afraid I ran out of time before I could properly review `conio.py` and `test_conio.py`, but the issues above should be enough to start with.

It looks like khorn has the most involvement with this code at the moment, so assigning back to them to respond to these issues.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_21"></a>|@khorn commented|
|-|-|

Fair warning: I didn't write any of this code to begin with.  I just unified the existing patches and made sure they worked with trunk.

That said, I'm happy to give this one a shot, as long as everyone else is happy to wait long enough for me to more fully understand the code. :)

I'm a little perplexed at how I should proceed though.  There are now two branches, and the newer one doesn't seem to have all (any?) of the changes in the patches for this ticket.  Should I be submitting patches against one of the branches?  Which branch?

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_22"></a>|@khorn commented|
|-|-|

Anyone?  Suggestions?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_23"></a>|@therve commented|
|-|-|

Unless I'm missing something, the latest branch should have all the changes submitted. Do you see anything missing?

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_24"></a>|@khorn commented|
|-|-|

OK, I think I know what happened now.  When I looked at the branches before, I was browsing the code in Trac, and several files were missing from the latest branch.  They all seem to be there now.  Maybe some sort of caching issue with Trac?

At any rate, seems fine now, my fault for not thinking of caching issues earlier.

Sorry. :)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_25"></a>|@khorn commented|
|-|-|

(In [28083]) Change line delimiter in doc/core/examples/stdin.py to work on Win32.

Refs: [#2157](https://github.com/twisted/twisted/issues/2157)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_26"></a>|@khorn commented|
|-|-|

(In [28085]) Address review comments:

 * Removed extraneous commented line from _pollingfile.py
 * fixed incorrect loop in _win32stdio.py
 * renamed win32Enabled to win32EventReactorInstalled
 * renamed conio.py to win32conio.py
 * fixed test_conio.py so that the tests don't blow up on non-win32 platforms

Refs: [#2157](https://github.com/twisted/twisted/issues/2157)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_27"></a>|@khorn removed owner|
|-|-|

Replying to [TimAllen](#note_20):

changes in [28085] address some of the review comments

>  1. There's a lot of "XXX" and "TODO" comments in this code. The coding standard says "Comments marked with XXX or TODO must contain a reference to the associated ticket". Hopefully the questions and unhandled conditions marked by these strings will be answered and handled before this branch is merged to trunk, and no tickets will need to be filed. :)

Let's see if we can fix up the other stuff first...

>  2. `t/i/_pollingfile.py`, line 239 in `_PollableReadPipe.checkWork()`: Commented-out code should be fixed if it ought to work, or deleted otherwise.

Removed the offending line.

>  3. `t/i/_win32stdio.py`, line 34 in `_PollableReadConsole.checkWork()`: Rather than explain the magic value 2000 as "80x25", why not just write it as "80 * 25"? Also, does this break if the console is a different size (such as 80×50)?

Well, it adds one more operation every time the loop runs, so maybe this was a premature optimization?

I currently have no idea if this breaks when/if the console size is different, but I don't see why it would.  For that matter I'm not sure why this value was chosen at all...

>  4. `t/i/_win32stdio.py`, line 36 in `_PollableReadConsole.checkWork()`: If you do get an unexpected `IOError`, it would be more helpful to raise it than to raise an unrelated `AssertionError`.

Are we sure that the AssertionError would be raised in all cases?  I'm not sure why it does what it's doing, but I also don't really know what the correct behavior should be, so...

>  5. `t/i/_win32stdio.py`, in `_PollableWriteConsole.checkWork()`: Variables `numBytesWritten` and `nBytesWritten` in the same method? There's got to be better names than these.

Probably so, but it follows the example of t.i._pollingfile.py (before this branch), which does the same thing in several places.

>  6. `t/i/_win32stdio.py`, line 90: Is there a reason that `ConsoleReader` doesn't have a leading underscore, where `_PollableReadConsole` (presumably exactly the same thing but for other reactors) does not?

>  7. `t/i/_win32stdio.py`, line 90: You have `ConsoleReader` for `win32eventreactor` and `_PollableReadConsole` otherwise; but `_PollableWriteConsole` works with all reactors. Why not drop `ConsoleReader` and just use `_PollableReadConsole` all the time, too?

>  8. `t/i/_win32stdio.py`, line 92 in `ConsoleReader.__init__()`: This is a terrible docstring. It doesn't describe what the class represents or how it's intended to be used.

>  9. `t/i/_win32stdio.py`, line 106 in `ConsoleReader.doRead()`: This looks exactly like `_PollableReadConsole.checkWork()`. Maybe `ConsoleReader` and `_PollableReadConsole` should inherit from some common base class or mixin? Otherwise, all the comments for `_PollableReadConsole.checkWork()` apply here too.

re: 6-9 -- I really am not sure what ConsoleReader was intended to do...if anyone has any ideas, I'd like to hear them.

>  10. `t/i/_win32stdio.py`, line 169 in `StandardIO.__init__()`: You say "for stdfd in range(0, 1, 2)", but I don't think that does what you intend. How about "for stdfd in (0, 1, 2)"?

Yikes!  Fixed as suggested.

Not sure why it worked before though...

>  11. `t/i/_win32stdio.py`, line 176 in `StandardIO.__init__()`: `win32Enabled` is an unfortunate name - it's quite possible to be using Win32 with the select reactor or IOCP reactor. How about calling it `eventReactorInstalled`?

Renamed to `win32EventReactorInstalled`.

>  12. `t/i/_win32stdio.py`, line 210 in `StandardIO.__init__()`: I feel it would be more readable to paste "`self._addPollableResource(self.stdin)`" into both places where it's required than to do the little if dance afterward.

I don't understand what is meant here. Can you clarify?

>  13. `t/i/conio.py`: Since this is just console IO for Windows rather than console IO in general, perhaps this module should be called `win32conio.py`?

Agreed.  Fixed.

However, it looks as though this may have been intended to grow into a platform-independent module.
See: http://twistedmatrix.com/trac/ticket/2157#comment:6 

???

>  14. `t/t/test_conio.py`: Why is this in `twisted/test/` and not `twisted/internet/test/`, since it's testing things in `twisted/internet/`?

I'm guessing the original author put it there because test_stdio is also in `twisted/test`.  Should it be moved?  Also, should it be renamed to something like `test_win32conio.py`?

>  15. `t/t/test_conio.py`: This test suite explodes horrible on non-Win32 platforms (or presumably Win32 platforms without the Win32 Extensions package installed) because it imports `win32console`. See `t/i/t/test_pollingfile.py` for an example of how to write platform-dependent tests that don't explode on other platforms.
>

I think this is fixed now, and the tests should be skipped on non-Win32.  Please test on non-Win32 platforms to make sure.


I've also noticed something a bit strange when running the tests.  Some of the tests in test_conio.py send output to the console which is running trial.  So it ends up sending out:

```

  ConInRawTestCase
    testFlush ...                                                          [OK]
    testRead ...                                                           [OK]
    testReadMultiple ...                                                   [OK]
    testReadWithBuffer ...                                                 [OK]
    testReadWithDelete ...                                                 [OK]
  ConInTestCase
    testBuffer ... hello
[OK]
    testDeleteBoundary ...w
                                                [OK]
    testDeleteFullBoundary ... hhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh
hhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh
hhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh
hhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh
hhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh
hhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh
w
                                            [OK]
    testFlush ... hello

```

Somehow I don't think is intentional. :)

I'm not sure if this is because of something trial is doing (or mis-assuming), or if it has to do with the code that writes to the console input buffer.

It seems to happen when WriteConsoleInput() is called.

The tests pass regardless...so is this really a problem with functionality?  I don't know.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_28"></a>|@jesstess set owner to @khorn|
|-|-|

Thanks for all this work, khorn!

[builds](http://buildbot.twistedmatrix.com/builders/documentation/builds/257)

Unfortunately there are both testing and markup issues.

If you're interested in running the buildbot tests yourself, a starting point is UsingBuildbot.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_29"></a>|@khorn commented|
|-|-|

(In [29541]) fix typo to skip tests properly on non-win32

refs: [#2157](https://github.com/twisted/twisted/issues/2157)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_30"></a>|@khorn commented|
|-|-|

(In [29542]) Docstring format fix

refs: [#2157](https://github.com/twisted/twisted/issues/2157)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_31"></a>|@khorn commented|
|-|-|

the only test that's failing on all the windows builders (aside from the glib2reactor and gtkreactor failures, which are unrelated)

```
[ERROR]: twisted.test.test_conio

Traceback (most recent call last):
  File "C:\twistedbot\winxp32-py2.6-select\Twisted\twisted\trial\runner.py", line 563, in loadPackage
    module = modinfo.load()
  File "C:\twistedbot\winxp32-py2.6-select\Twisted\twisted\python\modules.py", line 381, in load
    return self.pathEntry.pythonPath.moduleLoader(self.name)
  File "C:\twistedbot\winxp32-py2.6-select\Twisted\twisted\python\reflect.py", line 462, in namedAny
    topLevelPackage = _importAndCheckStack(trialname)
  File "C:\twistedbot\winxp32-py2.6-select\Twisted\twisted\python\reflect.py", line 398, in _importAndCheckStack
    return __import__(importName)
  File "C:\twistedbot\winxp32-py2.6-select\Twisted\twisted\test\test_conio.py", line 17, in <module>
    from twisted.internet import win32conio, _win32stdio as stdio
  File "C:\twistedbot\winxp32-py2.6-select\Twisted\twisted\internet\win32conio.py", line 337, in <module>
    stdin = ConIn(win32console.GetStdHandle(win32console.STD_INPUT_HANDLE))
  File "C:\twistedbot\winxp32-py2.6-select\Twisted\twisted\internet\win32conio.py", line 57, in __init__
    defaultMode = handle.GetConsoleMode()
pywintypes.error: (6, 'GetConsoleMode', 'The handle is invalid.')
```

I can't replicate this locally.  It looks like it's unable to get a proper handle to the console's std_in buffer.  Some possible causes (guessing):

1. Running under the buildbot makes it so that there's no handle to get.  How does buildbot run it's jobs?  In a separate process maybe? Or does something else already have a hold on the stdin buffer?
1. Some kind of mismatch between the version of pywin32 on the buildslaves and what's on my local machine?
1. Any other ideas?

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_32"></a>|@khorn commented|
|-|-|

(In [29543]) Branching to 'win-stdio-2157-3'

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_33"></a>|@khorn commented|
|-|-|

Merged forward and the tests still pass on my machine...still need to fix buildbot errors.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_34"></a>|@khorn commented|
|-|-|

(In [29544])  * Merge forward

   Refs: [#2157](https://github.com/twisted/twisted/issues/2157)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_35"></a>|@khorn commented|
|-|-|

After some research, it appears that buildslaves usually (always?) run as Windows Services on Win32.  This _might_ mean that the build user would need the "Allow service to interact with desktop" right in order to create a console.  Which _might_ have something to do with the above test failure.

Still can't replicate the failure locally.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_36"></a>|@khorn commented|
|-|-|

Success!

Finally replicated the error using:

```
>>> from subprocess import Popen
>>> outfile = open('outfile.txt', 'w')
>>> Popen([r'c:\python26\scripts\trial.bat', 'twisted.test.test_conio'], stdin=outfile)
```

I get the exact error above.  So now I guess it's time to figure out how to build a console or something in the test code...

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="gcc's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_38"></a>|gcc commented|
|-|-|

From MSDN:

```
GUI applications are initialized without a console. Console applications are initialized with a console, unless they are created as detached processes (by calling the CreateProcess function with the DETACHED_PROCESS flag).
```

I.e. your process probably doesn't have a console if you run it in the background or start it as a system service. Then all console-related Win32 APIs will fail.

You can call AllocConsole to get yourself a console. It might appear on the user's screen, at least temporarily, but that might be OK for a buildbot. I'm not sure how you could get any input into it in this state, so it might not be useful for testing, but at least you'd get past GetConsoleMode() failing.

**Comment by twisted-trac** [comment]
|<img alt="johnnypops's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_39"></a>|johnnypops commented|
|-|-|

I've been working on improved windows support for StandardIO() and found this branch to be a useful starting point. I've managed to get the (slightly modified) cftp.py and conch.py scripts working and solved the build-bot missing console problem. Improvements include being able to see the console cursor and scroll-back that doesn't keep 'pulling' you back to the active console line. I also tried the imap4client.py example program which worked with no changes.

I had to setup a buildbot-server for testing and have a windows 7 buildbot client (others to follow). I've tested my patches on XP/7/Linux and they seem to work OK. They are against the current trunk 38158.

Unfortunately, the patch is huge (90k) plus 3 new files (8k each) so I'm not sure exactly what to do with it.

For now I'm going to break it up into windows stdio support, tests for that support, conch/cftp changes, tests for that, and finally other unrelated issues I encountered whilst trying to get as many windows tests to pass as possible (6999).

Should I add them here when I'm ready?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_40"></a>|@exarkun commented|
|-|-|

Hi johnnypops.  I think it would be best if you raised this on the mailing list.  It would probably also be useful if you shared the code at the same time.  It's possible someone will be able to take a look at it and give you some ideas about the way to proceed which has the greatest chance of getting your code into Twisted quickly and easily.  Thanks!

**Comment by twisted-trac** [comment]
|<img alt="johnnypops's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_41"></a>|johnnypops commented|
|-|-|

I foolishly forgot to test with the stdiodemo.py and stdin.py examples mentioned in the original post. Attached fixed versions of win32conio.py and _win32stdio.py. 

stdin.py worked as-is but stdiodemo.py needs the delimiter in WebCheckerCommandProtocol() fixing:
```
class WebCheckerCommandProtocol(basic.LineReceiver):
    from os import linesep as delimiter

    def connectionMade(self):
```

It also revealed that connectionLost() wasn't being called correctly when using the console.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_42"></a>|@lvh commented|
|-|-|

This ticket isn't up for review, but it looks like perhaps it should be. Did you intend to put it up for review?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_44"></a>|@glyph commented|
|-|-|

I suspect that johnnypops may not be paying attention any more, so I'm going to attach the review keyword.  (If the author doesn't reply, please consider me the author on this now...)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_45"></a>|@glyph commented|
|-|-|

(In [42962]) Branching to win-stdio-2157-4.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_46"></a>|@glyph set owner to johnnypops|
|-|-|

johnnypops, thanks very much for your contribution.  I'm sorry it got lost in the shuffle.

I had a look at the attached patch and attempted to apply it to a branch.  It doesn't apply cleanly to trunk any more, and the patch that's there is missing a lot of stuff.  I couldn't really tell which files I was supposed to apply and which I wasn't.  It also randomly deleted a bunch of docstrings, changed whitespace to violate the coding standard in various places, and added a bunch of methods whose purpose wasn't clear.

All the new classes, `Channel`, `ChannelReadPipe`, `ChannelWritePipe`, `ChannelConsole`, `_PollableReadConsole`, and so on, need docstrings.

Also I don't quite understand tradeoffs involved in using `SetHandleInformation` vs. `DuplicateHandle`/`CloseHandle`, if that affects the way Twisted will behave at all.

I really hope someone will pick this back up, I'll try to review it more promptly in the future :).

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_47"></a>|@khorn set owner to @khorn|
|-|-|

I'll try to give this one another shot.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_48"></a>|@glyph commented|
|-|-|

Thanks khorn!  Looking forward to it! :)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_49"></a>|@khorn commented|
|-|-|

I've managed to get johnnypops' patches merged, but am having some trouble pushing them back up to SVN.  I'll try to get it pushed up tomorrow.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_50"></a>|@glyph commented|
|-|-|

Replying to [khorn](#note_49):
> I've managed to get johnnypops' patches merged, but am having some trouble pushing them back up to SVN.  I'll try to get it pushed up tomorrow.

What kinds of issues are you having?  When git takes a long time to do things with git-svn, sometimes it's a clue that it's preparing to do some enormously destructive operation that will mangle the whole repository, so I'm very keen to see the output of `git svn dcommit -n` or somesuch now :).

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_51"></a>|@khorn commented|
|-|-|

First I had an issue with git-svn finding my ssh binary (plink.exe in this case).  This was a matter of setting env vars correctly so it would find it, which sounds simpler than it actually was. :/

then I got something like this:

```
>git svn dcommit --dry-run
Committing to svn+ssh://svn.twistedmatrix.com/svn/Twisted/branches/win-stdio-2157-4 ...
diff-tree 2f0571134babb9902c7dc0976fde3913be50c705~1 2f0571134babb9902c7dc0976fde3913be50c705
```

which I _think_ means it should work fine (the hash is indeed the commit I want to push).  When I ran it without the --dry-run flag, though, it just sat there forever, giving me no feedback.  I figured it was up to no good, so I killed it after several minutes.

I'll investigate further presently...if you have any suggestions though I'd be glad to hear them.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_52"></a>|@khorn commented|
|-|-|

(In [43039]) Applied patches from johnnypops to recent branch

Refs: [#2157](https://github.com/twisted/twisted/issues/2157)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_53"></a>|@khorn commented|
|-|-|

Updated the latest branch with modified patches.  Tests seem to pass in my environment, and the stdin.py and stdiodemo.py examples work.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_54"></a>|@glyph commented|
|-|-|

Thanks again for looking at this, khorn.

Yes, that git-svn output looks correct.
